atc/exec/set_pipeline_step.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/concourse/concourse/atc/creds"
+	"github.com/concourse/concourse/atc/db"
+	"github.com/concourse/concourse/atc/db/lock"
+	"github.com/concourse/concourse/atc/exec/artifact"
+	"github.com/concourse/concourse/vars"
+)
+
+//go:generate counterfeiter . SetPipelineDelegate
+
+type SetPipelineDelegate interface {
+	BuildStepDelegate
+
+	Initializing(lager.Logger)
+	Starting(lager.Logger)
+	Finished(lager.Logger, ExitStatus, db.ConfigVersion, db.ConfigVersion)
+
+	FindTeam() (db.Team, bool, error)
+}
+
+// SetPipelineStep sets a pipeline to current team. If pipeline_name specified
+// is "self", then it will self set the current pipeline. This step takes pipeline
+// configure file and var files from some resource in the pipeline, like git.
+type SetPipelineStep struct {
+	planID      atc.PlanID
+	plan        atc.SetPipelinePlan
+	metadata    StepMetadata
+	delegate    SetPipelineDelegate
+	lockFactory lock.LockFactory
+	succeeded   bool
+}
+
+func NewSetPipelineStep(
+	planID atc.PlanID,
+	plan atc.SetPipelinePlan,
+	metadata StepMetadata,
+	delegate SetPipelineDelegate,
+	lockFactory lock.LockFactory,
+) Step {
+	return &SetPipelineStep{
+		planID:   planID,
+		plan:     plan,
+		metadata: metadata,
+		delegate:    delegate,
+		lockFactory: lockFactory,
+	}
+}
+
+func (step *SetPipelineStep) Run(ctx context.Context, state RunState) error {
+	logger := lagerctx.FromContext(ctx)
+	logger = logger.Session("set-pipeline-step", lager.Data{
+		"step-name": step.plan.Name,
+		"job-id":    step.metadata.JobID,
+	})
+
+	step.delegate.Initializing(logger)
+
+	variables := step.delegate.Variables()
+	params, err := creds.NewParams(variables, step.plan.Params).Evaluate()
+	if err != nil {
+		return err
+	}
+
+	spParams, err := atc.NewSetPipelineParams(params)
+	if err != nil {
+		return err
+	}
+
+	repository := state.Artifacts()
+	artifactSource := setPipelineArtifactSource{
+		ctx:    ctx,
+		logger: logger,
+		params: spParams,
+		repo:   repository,
+	}
+
+	atcConfig, err := artifactSource.FetchConfig()
+	if err != nil {
+		return err
+	}
+
+	step.delegate.Starting(logger)
+
+	stdout := step.delegate.Stdout()
+
+	if spParams.PipelineName == "self" {
+		spParams.PipelineName = step.metadata.PipelineName
+	}
+
+	team, found, err := step.delegate.FindTeam()
+	if err != nil {
+		logger.Error("failed-to-find-team", err)
+		return err
+	}
+	if !found {
+		err = fmt.Errorf("team not found")
+		logger.Error("failed-to-find-team", err)
+		return err
+	}
+
+	fromVersion := db.ConfigVersion(0)
+	pipeline, found, err := team.Pipeline(spParams.PipelineName)
+	if err != nil {
+		return nil
+	}
+
+	var existingConfig atc.Config
+	if !found {
+		existingConfig = atc.Config{}
+	} else {
+		fromVersion = pipeline.ConfigVersion()
+		existingConfig, err = pipeline.Config()
+		if err != nil {
+			return err
+		}
+	}
+
+	diffExists := existingConfig.Diff(stdout, *atcConfig)
+	if !diffExists {
+		logger.Debug("no-diff")
+
+		if spParams.FailWhenNoDiff {
```

Comment:
IMO steps should never fail in this case; they should just silently no-op. I wouldn't even have a flag for it, since this is something Concourse is pretty opinionated about (builds being idempotent). For example I might have one job that configures a bunch of pipelines all at once; it would be pretty common for all but one to be a no-op.